### DISCUSSION TITLE
Desktop Platform Support

### DISCUSSION THREAD

**Discussion by NathanaelA** [NativeScript/rfcs#11] [1. Ideas & Discussions]
One of the largest requested feature over NativeScript life has been Desktop support.
---

Bundyo created a basic prototype using [NodeGui](https://github.com/nodegui/nodegui) since this project actually support Linux, Mac and Windows and used QT for its display, and allows packing into a final executable.

I believe it is one of directions to proceed to create a method to create Gui apps in NativeScript for the Desktop.    If we can make a NS-Core app fully compatible, the other flavors should be able to work with minor to zero changes as all of them use the core pieces.   Unfortunately NodeGui is still early in its life cycle and might require changes to NodeGui.  

The other possibility it to use Tauri, or Neutralino it supports Linux, Mac, Windows native Executable; but offers some functionality that NodeGui does not, if we were to go the Tauri route, we could probably technically support pure HTML/web target for all flavors in addition.  

Tauri is rust based and its runtime is very small, making it a awesome candidate.
Neutralino is C++ based and its runtime is very small also.

## Relevant links prototype code: 
- https://github.com/bundyo/nativescript-platform-desktop from NS6 days, so it would have to be updated.

## Node Gui & Tauri
- [NodeGui](https://github.com/nodegui/nodegui) 
- [Tauri](https://github.com/tauri-apps/tauri)
- [Neutralino](https://neutralino.js.org/)

## 
Please add :+1: if you want to see this, not me to comments.   I would like to keep the discussion to debating which of the two directions to go, and/or people who actually want to help build this.

**Comment by darko-bw** [comment]
Hi @NathanaelA

Yes, it would be a massive feature to have. Completely agree with you.

**Comment by 7ammer** [comment]
I've wished for this for a while now! 😎👍
It would also be cool if nativescript was able to utilise mac catalyst.

**Reply by rigor789** [reply]
The runtime technically can (v8) - but we have not experimented with it... 

The required libs are built for maccatalyst: 
* https://github.com/NativeScript/ns-v8ios-runtime/tree/master/NativeScript/lib/x86_64-maccatalyst
* https://github.com/NativeScript/ns-v8ios-runtime/tree/master/NativeScript/lib/arm64-maccatalyst

**Reply by 7ammer** [reply]
Cool! I'll give this a try if my project allows it :)

**Comment by asharghi** [comment]
I bet this is a unpopular opinion, but I don't think we should support desktop platform for now. The NS team has already a lot on their plate with the mobile platforms. Updating Core for almost every major release of Android and iOS, making sure the CLI works with the newest Xcode version and MacOS, helping with all the plugins having deprecated code, keeping track on policy changes to App Store and Google Play that requires code changes, and so on. This is just some example of tasks that needs to be done for NS to survive, but I'm sure we want NS to thrive as well. I love NS, but the condition it is in today requires a lot of hacks and workarounds to make things work. Ironing out these things, and further develop for the mobile platform (like a working app sync https://github.com/NativeScript/rfcs/discussions/35) is in my opinion much more important. Then we can consider adding desktop platform support when it won't affect the mobile platforms negatively.

**Reply by buu700** [reply]
I brought this up in another thread, but a good middle ground could be just building web support in lieu of native desktop support, which would essentially act as a catch-all for any current or future unsupported platforms via Electron/Cordova.

**Reply by farfromrefug** [reply]
I agree 100% but i also think if someone wants to "join" the team and start working on this, they would be welcome to do so. This is how vue or svelte happened in N.

**Reply by vallemar** [reply]
I agree 100% @asharghi, a while ago I commented something like this in the general discord thread. I think that before something like this, NS needs to stabilize all the errors that you comment, webpack, core, debug, plugins ... I think stabilizing it is very important since as you say: tricks and errors usually do not tell us and new people arrive and as you see an error goes away. That said, I love NS and I want it to prosper and the most important thing now, I think it should be to solve everything. 

@farfromrefug I would love to join the team but it is very difficult, I do not have a deep knowledge of NS and much less of iOS, I think this is a problem for new collaborators.

That is my opinion, once everything is corrected, go ahead with the desktop

**Comment by rzr** [comment]
Even If desktop is not the priority, I would say that this can be valuable for many IoT devices running GNU/Linux.

Relate-to: https://github.com/NativeScript/NativeScript/issues/27#issuecomment-1023095347

**Comment by NathanaelA** [comment]
I've actually played with all three of those technology stacks I listed in my first message as possible directions to investigate and I don't believe any of them are the best way to proceed forward for a NS Desktop stack.    In fact I played with about 10 different stacks and have an entire blog series on it if you are at all curious starting here: https://fluentreports.com/blog/?p=1288) 

I do have a very early prototype of a NS-Desktop which I have a blog article about here: https://fluentreports.com/blog/?p=1422

**Reply by Jordan-Hall** [reply]
@NathanaelA  is this something thats being actively worked on?
